contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,304 @@
+    versionManager = _versionManager;
+
+    for (uint256 i = 0; i < _trustedProvisioners.length; i++) {
+      _addTrustedProvisioner(_trustedProvisioners[i]);
+    }
+
+    emit Setup();
+  }
+
+  function setPaused(bool _paused) external onlyOwner {
+    paused = _paused;
+    emit PausedToggled(_paused);
+  }
+
+  function getTrustedProvisioners() external view returns (address[] memory) {
+    return trustedProvisioners.values();
+  }
+
+  function _addTrustedProvisioner(address _token) internal {
+    trustedProvisioners.add(_token);
+    emit TrustedProvisionerAdded(_token);
+  }
+
+  function removeTrustedProvisioner(address _token) external onlyOwner {
+    trustedProvisioners.remove(_token);
+    emit TrustedProvisionerRemoved(_token);
+  }
+
+  function provisionPrepaidCard(address customer, bytes32 sku)
+    external
+    returns (bool)
+  {
+    require(!paused, "Contract is paused");
+    require(
+      trustedProvisioners.contains(msg.sender),
+      "Only trusted provisioners allowed"
+    );
+    require(asks[sku] > 0, "Can't provision SKU with 0 askPrice");
+
+    PrepaidCardManager prepaidCardManager = PrepaidCardManager(
+      prepaidCardManagerAddress
+    );
+
+    address token = skus[sku].issuingToken;
+    uint256 faceValue = skus[sku].faceValue;
+    string memory customizationDID = skus[sku].customizationDID;
+    uint256 priceToCreatePrepaidCard = prepaidCardManager.priceForFaceValue(
+      token,
+      faceValue
+    ); // in Wei
+
+    uint256[] memory issuingTokenAmounts = new uint256[](1);
+    uint256[] memory spendAmounts = new uint256[](1);
+
+    issuingTokenAmounts[0] = Exchange(exchangeAddress).convertFromSpend(
+      token,
+      faceValue
+    );
+
+    spendAmounts[0] = faceValue;
+
+    address issuer = skus[sku].issuer;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    IERC677(token).safeTransferAndCall(
+      prepaidCardManagerAddress,
+      priceToCreatePrepaidCard,
+      abi.encode(
+        customer,
+        issuingTokenAmounts,
+        spendAmounts,
+        customizationDID,
+        address(0), // marketAddress - we probably don't need it in this case?
+        issuer,
+        issuerSafe
+      )
+    );
+
+    balance[issuerSafe][token] -= priceToCreatePrepaidCard;
+
+    return true;
+  }
+
+  function getQuantity(bytes32 sku) public view returns (uint256) {
+    PrepaidCardManager prepaidCardManager = PrepaidCardManager(
+      prepaidCardManagerAddress
+    );
+
+    address token = skus[sku].issuingToken;
+    uint256 faceValue = skus[sku].faceValue;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    uint256 price = prepaidCardManager.priceForFaceValue(token, faceValue);
+
+    return balance[issuerSafe][token] / price;
+  }
+
+  function setAsk(
+    address issuer,
+    bytes32 sku,
+    uint256 askPrice
+  ) external returns (bool) {
+    require(skus[sku].issuer != address(0), "Non-existent SKU");
+    require(skus[sku].issuer == issuer, "SKU not owned by issuer");
```

Comment:
we also need a require (and a test) that prevents a random person from setting the ask. these will be safe txns, so the msg.sender should be the issuer's safe for the SKU that you are trying to set.

Reply:
Now that we added `onlyHandlers` here - does this suffice to satisfy this requirement? The handler will already check `require(issuer == owner, "only issuer can set market inventory");`

Reply:
Yes, I agree.